README.md
# Weather34 skin for WeeWX
Weather Station website skin with Live Data for WeeWX. This version is compatible with WeeWX 3.9.2 / Python 2.7 and WeeWX 4.x.x builds / Python 2.7 and Python 3.x. **Please note, some functionality will be lost if you are not using WeeWX 4.1.1 and its default extended database.** It is therefore strongly recommended that you start with an entirely new clean Python3 install of WeeWX 4.1.1.

Version W34-HC-IMJD-2.0.0

Packaged for installation using its own unique installer.

This repository contains the WeeWX version of Brian Underdown's Weather34 website template set. Brian's main website is https://weather34.com/homeweatherstation/index.html In January, 2019, Brian asked others to take over the distribution/maintainence of his design whilst he concentrated on development only for MB NanoSD, called Weather34 MB-SMART. This WeeWX version of the template is actively developed by Ian Millard and Jerry Dietrich.

This version requires WeeWX version 3.9.2 or later software. WeeWX is available at http://weewx.com

This version is designed explicitly to harness the powerful WeeWX database to generate the weather data charts and statistical data. It was originally built on the current MB-UB40-IHVN which is now maintained by Lightmaster (Meteobridge-Weather34-Template). Although we maintain a similar look and feel with MB-UB40-IHVN,the weewx-Weather34 now has many unique features.

This version is assembled as an install package and uses its own unique utility, w34_installer.py, to install. This greatly simplifies the installation process from that of previous versions. Depending on your own WeeWX setup, minor edits may be required to be made to weewx.conf and Weather34 skin.conf files. Please see the Weather34 skin Installation Guide for detailed instructions.

# What's New in this Version W34-HC-IMJD-2.0.0

1. New air quality modules (US/Global AQI and UK/DAQI) for the top row and an updated main air quality module. These all rely on you using the new default extended datbase schema which includes records for particle concentrations. You must have an air quality particle detector which is able to populate these fields. There are Highcharts setup to plot particle data concentrations. These new modules are selectable from the settings page.

2. Cloud Coverage. Cloud coverage is displayed in the current conditions module expressed as % and oktas(eighths). There are Highcharts set up to plot % coverage. The cloud coverage is measured from satellite images and is adapted from a method proposed by David Marshall.

3. For Geographical European users only, Meteoalarm weather alerts. The warnings are flagged in the advisory top row module which has a hot link to a pop-up window showing the full detail. Up to four different warnings for a location can be displayed in the pop-up window. Again thanks to David Marshall for this proposal. All settings are contained in the web services section of the setup page. I am keen to expand this idea for other parts of the globe. As I am not familiar with warning/advisory services in othe parts of the world, I am open to any users' contribution of working code which fits in with the Weather34 design.

4. Connections to a remote server. Thanks to Jerry Dietrich, a more robust method of connection to a remote server has been written into the weather34.py code. Please see separate instructions for setting up a remote server.

5. Additional Highcharts from Jerry Dietrich include twin wind rose, wind Hays chart, cloud coverage, air particle concentrations, improved radial charts. 


# Demo

A live example of Weather34 WeeWX skin can be seen at https://claydonsweather.org.uk

# Weather34 Historic Timeline of design 2014-2019 
https://weather34.com/homeweatherstation/weather34timeline/weather34timeline.html

# Template Screenshots

**Dark Theme**

![image](https://user-images.githubusercontent.com/18438654/86633765-fb60a200-bfc8-11ea-99dc-f8dc8de56e8c.png)

**Light Theme**

![image](https://user-images.githubusercontent.com/18438654/86635273-c8b7a900-bfca-11ea-9efd-76962364c2fd.png)

**Pop Up Chart - Dark Theme**

![image](https://user-images.githubusercontent.com/18438654/86635575-277d2280-bfcb-11ea-9452-e7acd7f44e62.png)

**Pop Up Chart - Light Theme

![image](https://user-images.githubusercontent.com/18438654/86636143-dae61700-bfcb-11ea-966d-4db7b487033f.png)

**Pop Up AQI Info**

![image](https://user-images.githubusercontent.com/18438654/86636425-2c8ea180-bfcc-11ea-957d-72c1180f44fb.png)

**Pop UP Alerts**

![image](https://user-images.githubusercontent.com/18438654/86637661-da4e8000-bfcd-11ea-8df6-1f872a6bb6ed.png)

**Radial Charts**

![Untitled](https://user-images.githubusercontent.com/18438654/86875356-34d50280-c0da-11ea-9b39-33acd451eb46.png)

**Hays Charts**

![image](https://user-images.githubusercontent.com/18438654/86639506-e25bef00-bfd0-11ea-9da3-043d2a640b46.png)


# Setup

**If you are upgrading from a previous version of this skin, due to the significant changes, you must recreate your settings!.php file by running the full install process. Just copying your previous settings1.php will cause errors which are then difficult to resolve.**

Follow the instructions in the 'installation guide' (INSTALLATION_GUIDE.md) to install the template.
Browse to http://your/path/to/weewx/weather34/easyW34skinSetup.php
There is no initial password when the page prompts the first time -- just press LOGIN to enter the page.
IMPORTANT set a password in the screen for future use -- your browser can remember it. This will make future updates    reasonably secure so only you can do the updates to the configuration.
Make setting entries in the easyweathersetup.php page and SAVE. The next time you run it, use the password you set in the step above.
Repeat running easyW34skinSetup.php until the main screen appears as you like it.

# License

Copyright (c) 2016-2019 by Brian Underdown (https://weather34.com)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Template”), to deal in the Template without restriction, including without limitation the rights to, can use, can not copy without prior permission, can modify for personal use, can use and publish for personal use ,can not distribute without prior permission, can not sublicense without prior permission, and can not sell copies of the Template, and subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Template.

THE TEMPLATE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE TEMPLATE OR THE USE OR OTHER DEALINGS IN THE TEMPLATE.

Attribution-NonCommercial 4.0 International based on a work at https://weather34.com/homeweatherstation
Non-weeWX versions Available

# An Excerpt from Meteobridge-Weather34-Template MB-UB40-RRW README.md
Github repository for the Meteobridge version of the original weather34 template 
Weather Template for Meteobridge users only 

# This work is not permitted to be used in any other versions without prior permission unless listed below 
# Permission is granted for use in Cumulus version maintained by Ken True 
# Permission is granted for use in Weewx version maintained by Ian Millard

*This work means CSS/SVG/PHP .

Meteobridge Version available via download maintained by Lightmaster (https://github.com/lightmaster/Meteobridge-Weather34-Template)

Cumulus Version available via download and more info supported by Ken True ( https://github.com/ktrue/CU-HWS ). This version now also supports WeeWX and WeatherCat

# Credits for this version

Apart from Brian Underdown without him, this template would never exist, I would aslo like to acknowledge the following people: -

Jerry Dietrich for his massive contribution in converting my wild ideas into reality, putting me straight on my mediocre coding skills and having the patience of a dozen saints.

Gary Roderick for his original coding of Highcharts for WeeWX.

William Bailey aka Lightmaster who maitains the MB version of Weather34. Incredibly helpful and always on the end of Telegram when I need a chat, night or day.

Ken True for sharing files and who makes my job of maintaining the WeeWX version so much easier.

Thomas Sosio for his invaluable contribution in producing the Meteobridge lookup code to translate WeeWX database output.

David Marshall for contributing technical knowledge and solutions to create some original .tmpl files and alternative solutions for weather alerts.

Taylormia for contributing his excellent setup example for instances where WeeWX and server/template are remote to each other.

Gary Portellas for a helpful suggestion to further simplify the installation process.

Drealine for his recent work on extending the French translations.

All those unamed people who have helped me with testing updates.

The creators of WeeWX without which there would be no point.

# Alternative versions

# Meteobridge
you can find the Meteobridge version maintained by Lightmaster (William) at 
https://github.com/weather34/Meteobridge-Weather34-Template34/

# Cumulus
you can find the Cumulus version maintained by Ken True (Saratoga) at 
https://github.com/ktrue/CU-HWS

# Weewx
you can find a Weewx version maintained by Ian Millard is also avaialble via 
https://github.com/steepleian/weewx-Weather34

# Weatherflow
you can find a Weatherflow version not maintained but fully tested as of May 5th 2019 
https://github.com/weather34/Weather34-Weatherflow


# This work is licensed under a Creative Commons Attribution-NonCommercial-NoDerivatives 4.0 International License.
http://creativecommons.org/licenses/by-nc-nd/4.0/

# Credits and thanks to the contributors who made the original version of Weather34 possible between 2015 and 2019.

 Erik M Madsen for language idea and initial script
 
 Paul @komoka weather in Canada for continous support and testing 
 
 Josep for Spanish/Catalan language translation and for many ideas and refinements
 
 Pascal Catte French translation and ideas fowarded 
 
 Steve the developer of Cumulus for support and providing a platform to resolve issues 
 
 Mats Ahlklo Swedish translation and his work on using Davis weatherlink 
 
 Betejuice (Cumulus Forum) for providing a solution for meteor shower listings 
 
 Ken True (Saratoga) for kindly granting permission allowing use of many scripts he developed which gave inspiration and ideas  though not used today it was the inspiration that allowed to do something more suited to the design. 
 
 Eric Rechlin Special thanks for originally creating the theme switching and extensive work on metrics/non metrics
 
 Boris at smartbedded (meteobridge) for ongoing support and upkeep of meteobridge 
 
 Wim van der Kuil for the original meteobridge script 
 
 David St John at weatherflow for providing hardware for testing and his non bias logical views 
 
 Paul Wilman , Tina Thomas, Vaggos , Chuck M , Aaron Gersztoff , Ian Millard, and many many more for continous constructive supportive feedback .. 


CHANGE_LOG.md
# Change Log

**Version W34-HC-IMJD-2.0.0**

1. Changes for WeeWX 4.0 new extended database schema
2. Protection code for empty strings
3. Added energy to lightning module display
4. CSS additions in preparation for future planned air quality module options
5. Updated settings.php to next version code **W34-HC-IMJD-2.0.0**
6. Added lightning chart links to index.php
7. Code cleanup lightning variables & css tweaks for lightning
8. Added lightning monthly chart
9. Add missing y axis label for lightning
10. Additional lightning label changes
11. Readme up issue
12. Removed duplicated documents in www folder
13. Adjust location of miles widget
14. Reduce race condition on update section
15. wind_gust_60min missing and wind conversions in the wrong file
16. Cloud height display reversed
17. CSS adjustment for airqualitymodule.php
18. Corrected issues with RT windrose chart
19. Further air quality module CSS adjustments
20. Air Quality Module updated
21. AQI info pop up added (aqipopup.php, aqi_info.php)
22. World AQI Top Row Module added (worldaqitop.php)
23. UK AQI Top Row added (daqitop.php)
24. Added cloudbase data to archivedata.php.tmpl
25. w34CombineData.php now uses WeeWX cloud base data 26 Fixed cloud base unit conversion in currentconditions.w34.php
26. Use lasttime for lightning strike
27. Created airqualitydata.php.tmpl
28. Added [[[AIRQUALITYDATA]]] to Weather34 skin.conf
29. Dual windrose plots on one iframe
30. Day plots for multiple plots per page
31. Cloudbase calculation works on US units database also
32. Month/Year Windrose Gust use GustSpeed from db
33. Windrose Gust tweaks and using 10min gust for display
34. Fix chk_lightning_count option
35. Remove unused config entry
36. Weewx IP in realtime.txt for day plots
37. Remove weewxbinpath from plots_config.js
38. Change to make lightning time ago correct
39. Round 2 Day Plots communication
40. Use 10min dir avg and 10min gust avg for rose
41. Config option to disable day plots
42. New Cloud Cover percent display
43. New Cloud Cover Chart
44. Misc changes for Cloud Cover and chmod for sat images
45. Added cloud cover unit to settings.php
46. Renamed RemoteWebServer and realtime.txt
47. Added text to cloud cover chart
48. Added additional hidden fields for new hidden variables to ensure they are not removed from settings1.php when settings page is saved.
49. Remove scroll bar from temp almanac
50. Corrected path in Weather34CloudCover stanza in the install config files
51. Added % cloud cover chart link to current conditions module
52. Rsync w34realtime.txt to remote web servers
53. Rsync of json directory is now possible
54. Remove duplicate filename config entries in W34 sections
55. Added green, yellow, orange and red advisory alert triangles to shared.php
56. Inserted space in cloud cover link index.php
57. Fixed white plot background for cloud cover chart
58. Buttons styles
59. W34 WebServer always enabled remove config section from weewx conf
60. Send webserver ip address on index.php page load
61. currentconditionsw34.php adjusted layout
62. Support multiple remote web servers for one weewx machine
63. Auto theme for dark and light theme based on sunrise & sunset
64. Added delete of config entries to w34 installer
65. Added advisory_extended.php
66. Added alert fields to easyW34skinSetup.php
67. $lat and $lon added to auto theme index.php
68. settings1.default.php additional variables for Meteoalarm added
69. Additional updates to easyW34skinSetup.php for Meteoalarm filelds
70. Hayes wind speed chart and range selector for radial charts
71. index.php edited module links
72. Protection code around lightningSum_json
73. Protection code for degrees to compass value
74. Cloud Cover Radial chart added
75. readme.md Updated screen shots
76. Updated radial chart screen shot
77. Added polar chart for tempallplot
78. Use either weewx 3.x or 4.x logging
79. Allow polar charts to use different dates using input selector
80. User is now prompted with the Python Version the install will be run
81. Re-worked cloud cover
82. Added night and day icon selection to cloud conditions
83. easyW34skinSetup.php MB platform not required
84. Added $theme1 and $al_ip to settings1.default.php
85. Added missing cloud cover variable to w34CombinedData.php
86. Current conditions module Icons and descriptions now match
87. Addition of oktas to cloud coverage in current conditions module
88. Readme and installation instructions updated
89. Updated weather34.py, change for time out in P3
90. Added w34_reports to w34Highcharts folder and updated plot_config.js accordingly
91. Rolled back 89. to previous state as it caused chart click-through to hang on final click
92. weather34.py corrected hard coded path
93. w34CombinedData.php try catch around loading archivedata.php 
94. Fix for Highcharts update.
95. Corrected font weight WU large forecast in light theme


www/settings.php
<?php 
include('settings1.php');error_reporting(0); 
	####################################################################################################
	#	HOME WEATHER STATION TEMPLATE SETUP please set up and check thoroughly                         #
	#	CREATED FOR HOMEWEATHERSTATION TEMPLATE at https://weather34.com/homeweatherstation/index.html # 
	# 	   FILE= SETTINGS.PHP                                                                          #
	# 	                                                                                               #
	# 	WEATHER34 HOMEWEATHER METEOBRIDGE MYSQL                                                        #
	#   https://www.weather34.com 	                                                                   #
	####################################################################################################


###########################################################################
# THE SETIINGS BELOW MUST BE LEFT UNTOUCHED UNLESS YOU REALLY NEED  #######
# TO MAKE THE CHANGES HERE TO MAKE IT WORK ON YOUR SERVER// ###############
# TAKE THE TIME TO STUDY THEM DONT TAKE IT FOR GRANTED      ############### 
# USE THE EASY SETUP PANEL TO MAKE CHANGES IN THE TEMPLATE  ###############
###########################################################################

$TZconf = $TZ;// PLEASE LEAVE it is fixed 
$api2= 'xxx'; // optional Weather Underground api key developer key '1234567890'
$rise_zenith = 90+ 40/60; // try 50/60 or something/60 until it matches correctly to your sunrise .this allows you to fine tune the sunrise
$set_zenith = 90+ 36/60;  // try 50/60 or something/60 until it matches correctly to your sunset .this allows you to fine tune the sunset
$forecastlocation = $stationlocation; // 
$version            = $livedataFormat; // template version and type of source: Clientraw, MeteoBridge, Cumulus, etc (for display only)
$emailform          = $email; // PLEASE LEAVE FIXED 
$timeFormatShort    = 'H:i'; // time format to use when showing times ('g:i a' for '5:34 pm'; 'H:i' for '09:28')
$showFeelsLike      = true; // whether to always show either the heat index (when temp > 80F/27C) or real feel (when temp between 50F/10C and 80F/27C) even when no concern
$lightLeft          = true; // shows amount of light/darkness left rather than total amount per day in the moon display
$theme              = $theme1;
$charttheme         = $theme;

####################################################################################################
// Refresh Data Main Page  //																	   #
// Automatic refresh times (in seconds) of each panel on the main dashboard						   #
####################################################################################################
$copyYear = 2015;
$curYear = date('Y');
$copyrightcredit='&copy; weather34.com original CSS/SVG/PHP '.$copyYear . (($copyYear != $curYear) ? '-' . $curYear : 'Copyright');
$indoorRefresh      = 600; // 10 minutes
$advisoryRefresh    = 60; // 60 seconds
$tempRefresh        = 60; // 60 seconds
$skyRefresh         = 600; // 10 minutes
$windSpeedRefresh   = 04; // 4 seconds
$baroRefresh        = 300; // 5 minutes
$windDirRefresh     = 04; // 4 seconds
$solarRefresh       = 600; // 10 minutes
$moonRefresh        = 1800; // 30 minutes 
$rainRefresh        = 60; // 60 seconds 
$p12Refresh         = 600; // 10 minutes        Position 12 refresh (if not webcam) 
$p13Refresh         = 600; // 10 minutes        Last Postion refresh (if not webcam)
$moonphaseRefresh   = 600; // 10 minutes
$eqRefresh          = 600; // 10 minutes 
$forecastRefresh    = 900; // 15 minutes 
$notificationcenterRefresh    = 300; // 5 minutes set refresh time for notification of earthquakes default 5 minutes
$wuupdate           = 1800; // 30 minutes 
$wfupdate           = 60; // weatherflow api 60 seconds
$camRefresh         = 60; // 1 minute refresh on camera popup
// Thresholds for warnings or notifications
$notifyDistEQ       = 300; // miles if $windunit is mph otherwise km
$notifyEQMag        = 5; // set this to minimum magnitude of earthquake for notifications, typically 5 or 6, or you can set to 12 to switch off if not required
$notifyWindGust     = 22; // knots; 22 knots, which is 40.7 km/h, 35 = 40.2 mph, 40 = 20.5  m/s
$notifyHeatIndex    = 30; // degrees C, default is 30 celsius which is 86 degrees F 
$notifyWindchill    = 3; // degrees C, default is 10 celsius which is 50 degrees F
$notifyFeelslike    = 3; // degrees C, default is 10 celsius which is 50 degrees F
$notifyDewpointcold  = 3; // degrees C, default is 8 celsius 46.4 degrees F OR WHATEVER MAKES YOU FEEL COLDER
$notifyRainRate     = 7.62; // millimeters, default is 10 which is 0.4 in"
$notifyDewPoint     = 21; // dewpoint comfortable zone alert 22 degrees C or 71.6 degrees F OR WHATEVER MAKES YOU FEEL UNCOMFORTABLY HOT
$notifyFlooding     = 25.4; // dewpoint comfortable zone alert 22 degrees C or 71.6 degrees F OR WHATEVER MAKES YOU FEEL UNCOMFORTABLY HOT

####################################################################################################
// Probably won't have to change anything past this line IF SO BE VER VERY CAREFUL!!!!!!		   #
####################################################################################################
$scriptcredits     = "Original CSS/SVG ICONS/PHP scripts by <a href='https://weather34.com' title='weather34.com' target='_blank'>weather34.com 2015 - ".date('Y').""; // for modules
$creditsEnabled     = "true"; // for chart pages only
$creditsmysql       = "Weather Data Recorded from $stationName Database"; //for chart pages only
$wucredits          = "Data provided by Weather Underground"; // for chart pages only
$creditschart 		= "CanvasJs.com v2.3.1 GA (CC BY-NC 3.0) Non-Commercial-Version";
$creditsURL         = ""; // for chart pages only
$credits            = "Data Supplied via Weather Underground"; // for chart pages only
$templateinfo       = ''; // template information page
$templateversion    = 'W34-HC-<maxblue>IMJD</maxblue>-2.0.0';
$software    = 'WeeWX <span>Hardware</span> Users';$designedfor='<br>For WeeWX Users';$chartversion='(WUDATACHARTS)';$somethinggoeshere ='d4586dec-e7a2-47ae-99b6-25527b2563c9';$chartversionmysql  =  '(DATACHARTS WeeWX Database Version)';
if (array_key_exists('theme', $_GET) && ($_GET['theme'] == 'dark' || $_GET['theme'] == 'light')) {
  SetCookie('theme', $_GET['theme'], time() + 15552000);
  $theme = $_GET['theme'];
} else if (array_key_exists('theme', $_COOKIE) && ($_COOKIE['theme'] == 'dark' || $_COOKIE['theme'] == 'light')) {
  $theme = $_COOKIE['theme'];
}
$units = "";
if (array_key_exists('units', $_COOKIE)) {
  $units = $_COOKIE['units'];
}
ini_set('session.use_cookies', '0');
if (array_key_exists('units', $_GET) && ($_GET['units'] == 'default' || $_GET['units'] == '')) {
  SetCookie('units', $_GET['units'], time() - 86400); //86400 = 1 day, negative time erases cookie
  $units = '';
} else if (array_key_exists('units', $_GET) && ($_GET['units'] == 'us' || $_GET['units'] == 'uk' || $_GET['units'] == 'metric' || $_GET['units'] == 'scandinavia')) {
  SetCookie('units', $_GET['units'], time() + 15552000);
  $units = $_GET['units'];
}
if ($units == 'uk') {
  $cloudcoverunit   = '%';	
  $windunit         = 'mph';
  $tempunit         = 'C';
  $rainunit         = 'mm';
  $pressureunit     = "hPa";
  $windconv         = "0.621371";
  $rainfallconv     = '10';
  $pressureinterval = "0.5";
  $rainfallconvmm   = '10';
} else if ($units == 'scandinavia') {
  $windunit         = 'm/s';
  $tempunit         = 'C';
  $rainunit         = 'mm';
  $pressureunit     = "hPa";
  $windconv         = "0.277778";
  $rainfallconv     = '10';
  $pressureinterval = "0.5";
  $rainfallconvmm   = '10';
} else if ($units == 'metric') {
  $windunit         = 'km/h';
  $tempunit         = 'C';
  $rainunit         = 'mm';
  $pressureunit     = "hPa";
  $windconv         = "1";
  $rainfallconv     = '10';
  $pressureinterval = "0.5";
  $rainfallconvmm   = '10';
} else if ($units == 'us') {
  $windunit         = 'mph';
  $tempunit         = 'F';
  $rainunit         = 'in';
  $pressureunit     = "inHg";
  $windconv         = "1";
  $rainfallconv     = '1';
  $pressureinterval = "0.5";
  $rainfallconvmm   = '1';
}
?>
